File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/domain/interactor/location/GetLocationUpdatesUseCase.java
package com.lattis.ellipse.domain.interactor.location;

import com.lattis.ellipse.domain.executor.PostExecutionThread;
import com.lattis.ellipse.domain.executor.ThreadExecutor;
import com.lattis.ellipse.domain.interactor.UseCase;
import com.lattis.ellipse.domain.model.Location;
import com.lattis.ellipse.domain.repository.LocationRepository;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;

/**
 * Created by raverat on 2/23/17.
 */

public class GetLocationUpdatesUseCase extends UseCase<Location> {

    private LocationRepository locationRepository;

    @Inject
    protected GetLocationUpdatesUseCase(ThreadExecutor threadExecutor,
                                        PostExecutionThread postExecutionThread,
                                        LocationRepository locationRepository) {
        super(threadExecutor, postExecutionThread);
        this.locationRepository = locationRepository;
    }

    @Override
    protected Observable<Location> buildUseCaseObservable() {
        return this.locationRepository.getLocationUpdates();
    }

    @Override
    public Disposable execute(DisposableObserver<Location> UseCaseSubscriber) {
        return this.buildUseCaseObservable()
                .subscribeWith(UseCaseSubscriber);
    }


}

File: velolabs_repos/lattis_android_v2/domain/src/main/java/com/lattis/domain/usecase/location/GetLocationUpdatesUseCase.kt
package com.lattis.domain.usecase.location

import com.lattis.domain.executor.PostExecutionThread
import com.lattis.domain.executor.ThreadExecutor
import com.lattis.domain.repository.LocationRepository
import com.lattis.domain.usecase.base.UseCase
import com.lattis.domain.models.Location
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.observers.DisposableObserver
import javax.inject.Inject


/**
 * Created by raverat on 2/23/17.
 */

class GetLocationUpdatesUseCase @Inject
constructor(threadExecutor: ThreadExecutor,
            postExecutionThread: PostExecutionThread,
            private val locationRepository: LocationRepository
) : UseCase<Location>(threadExecutor, postExecutionThread) {


    var freshLoctionData =false

    fun withFreshLocationData(freshLoctionData:Boolean):GetLocationUpdatesUseCase{
        this.freshLoctionData=freshLoctionData
        return this
    }


    override fun buildUseCaseObservable(): Observable<Location> {
        return this.locationRepository.getLocationUpdates(freshLoctionData)
    }

    override fun execute(UseCaseSubscriber: DisposableObserver<Location>): Disposable {
        return this.buildUseCaseObservable()
                .subscribeWith(UseCaseSubscriber)
    }


}
